sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteGroupingAnalyticsWithoutKeys.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, AttributeReference, AttributeSet, Literal, NamedExpression, VirtualColumn}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * In grouping analytics (e.g., `GROUPING SETS`) with an empty grouping set, we need to follow
+ * the non-key aggregate semantics; it generates a single row having an initial aggregated value
+ * (e.g., 0 for `COUNT`) from an empty input.
+ * For example, a query below must return a row (NULL, NULL, 0) from
+ * an empty input (`empty_input`):
+ *
+ * {{{
+ *   SELECT k1, k2, COUNT(v)
+ *     FROM empty_input t(k1, k2, v)
+ *     GROUP BY GROUPING SETS ((), (k1, k2))
+ * }}}
+ *
+ * To comply with the semantics, this rule rewrites the following (pseudo) resolved logical plan
+ * as an union query of aggregates with/without keys as follows:
+ *
+ * {{{
+ *   Aggregate(
+ *     groupingExprs = ['k1, 'k2, 'spark_grouping_id],
+ *     aggregateExprs = ['k1, 'k2, COUNT('v)],
+ *     output = ['k1, 'k2, 'cnt],
+ *     child = Expand(
+ *       projections = [('k1, 'k2, 'v, 0), (null, null, 'v, 3)],
+ *       output = ['k1, 'k2, 'v, 'spark_grouping_id]),
+ *       child = LocalTableScan [...]))
+ * }}}
+ *
+ * It is transformed into an union plan below:
+ *
+ * {{{
+ *   Union([
+ *     Aggregate(
+ *       groupingExprs = ['k1, 'k2, 'spark_grouping_id],
+ *       aggregateExprs = ['k1, 'k2, COUNT('v)],
+ *       output = ['k1, 'k2, 'cnt],
+ *       child = Expand(
+ *         projections = [('k1, 'k2, 'v, 0)],
+ *         output = ['k1, 'k2, 'v, 'spark_grouping_id]),
+ *         child = LocalTableScan [...])),
+ *     Aggregate(
+ *       groupingExprs = [],
+ *       aggregateExprs = [null, null, COUNT('v)],
+ *       output = ['k1, 'k2, 'cnt],
+ *       child = Project(
+ *         projectList = ['v],
+ *         child = LocalTableScan [...]))]
```

Comment:
For non-empty input case, does this mean we will re-read the input table multi-times?

Reply:
To run additional Aggregate multi-times sounds also a burden on performance.

Reply:
Yes. That's true in AS-IS PR. And, that could be a reason to drop this PR and JIRA completely.

Reply:
@viirya Yea, currently yes.

> And, that could be a reason to drop this PR and JIRA completely.


@dongjoon-hyun Do you think it is ok to keep the current behaviour?

Reply:
We have many open correctness or data-loss issues. This is one of them. Nobody thinks that is ok, but they are still open (or abandoned.)
> Do you think it is ok to keep the current behavior?

For this specific issue, although Presto and PostgreSQL has a way, if Apache Spark community cannot find a proper solution for us, it's difficult to land to master (even in 3.0.0). Consequently, there is no need to backport.

Reply:
Yea, dropping this pr is ok, but I'm not sure that its ok to drop the jira. Since the current behaviour seems to be illegal in in terms of the relational theory semantics, I think the should keep open.

Reply:
Sure. I mean "It could be", not "It should be".

For the following, you and I explained that reason and most people seems to agree that this is a correctness issue now. However, we still didn't reach an agreement on that Apache Spark need to have this like Presto and PostgreSQL.
> Since the current behaviour seems to be illegal in in terms of the relational theory semantics

There are many ways when a JIRA is closed. Sometimes, it's closed as 'Later'. So, sometimes, it's closed as 'Won't Fix'. More frequently, it's just open for a long time and closed as 'Incomplete' in a bulk close.

Reply:
ah, ok. Thanks for the explanation~